lib/controllers/run-controller.ts

```diff
@@ -245,18 +248,27 @@ export class RunController extends EventEmitter implements IRunController {
 
 			try {
 				let packageFilePath: string = null;
-				const shouldBuild = prepareResultData.hasNativeChanges || await this.$buildController.shouldBuild(buildData);
-				if (shouldBuild) {
-					packageFilePath = await deviceDescriptor.buildAction();
+
+				// Case where we have three devices attached, a change that requires build is found,
+				// we'll rebuild the app only for the first device, but we should install new package on all three devices.
+				if (this.rebuiltInformation[platformData.platformNameLowerCase] && (this.$mobileHelper.isAndroidPlatform(platformData.platformNameLowerCase) || this.rebuiltInformation[platformData.platformNameLowerCase].isEmulator === device.isEmulator)) {
```

Comment:
Check what happens if the builds are in the order iOS emulator, iOS device, iOS emulator.

Reply:
We discussed it personally and I'll handle this in another PR.